examples/Simple/Quicksort.hs

```diff
@@ -0,0 +1,61 @@
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Simple.Quicksort
+  ( testQuicksort
+  , quickSort
+  )
+where
+
+import GHC.Stack
+import qualified Data.Array.Mutable.Linear as Array
+import Data.Array.Mutable.Linear (Array)
+import Data.Unrestricted.Linear
+import Prelude.Linear hiding (partition)
+
+
+-- # Testing
+-------------------------------------------------------------------------------
+
+testQuicksort :: [Int] -> Bool
+testQuicksort xs = sort xs == quickSort xs
+
+
+-- # Quicksort
+-------------------------------------------------------------------------------
+
+quickSort :: [Int] -> [Int]
+quickSort xs = unur $ Array.fromList xs $ Array.toList . arrQuicksort
+
+arrQuicksort :: Array Int %1-> Array Int
+arrQuicksort arr = Array.size arr & \case
+  (Ur len, arr') -> go 0 (len-1) arr'
+
+go :: Int -> Int -> Array Int %1-> Array Int
+go lo hi arr = case lo >= hi of
+  True -> arr
+  False -> Array.read arr lo & \case
+    (Ur pivot, arr0) -> partition arr0 pivot lo hi & \case
+      (arr1, Ur ix) -> swap arr1 lo ix & \case
+        arr2 -> go lo (ix-1) arr2 & \case
+          arr3 -> go (ix+1) hi arr3
+
+partition :: Array Int %1-> Int -> Int -> Int -> (Array Int, Ur Int)
+partition arr pivot lx rx
+  | (rx < lx) = (arr, Ur (lx-1))
+  | otherwise = Array.read arr lx & \case
+      (Ur lVal, arr1) -> Array.read arr1 rx & \case
+        (Ur rVal, arr2) -> case (lVal <= pivot, pivot < rVal) of
+          (True, True) -> partition arr2 pivot (lx+1) (rx-1)
+          (True, False) -> partition arr2 pivot (lx+1) rx
+          (False, True) -> partition arr2 pivot lx (rx-1)
+          (False, False) -> swap arr2 lx rx & \case
+            arr3 -> partition arr3 pivot (lx+1) (rx-1)
+
+swap :: HasCallStack => Array Int %1-> Int -> Int -> Array Int
+swap arr i j = Array.read arr i & \case
+  (Ur ival, arr1) -> Array.read arr1 j & \case
```

Comment:
`LambdaCase`'s on this file seems unnecessary.

Reply:
Yes, true. I will keep the lambda notation instead of just using currying. I think it's more readable as we pass the result of the left of the `&` and bind it to the argument of the lambda, which is the `x` in `\x -> ...`.